runner/main.go

```diff
@@ -0,0 +1,269 @@
+	runner := &testRunner{
+		m:                m,
+		exitOnError:      exitOnError,
+		failRetriesCount: failRetriesCount,
+	}
+	runner.init()
+	return runner.Run()
+}
+
+// Gets the test name
+func GetTestName(name string) string {
+	if runnerRegexName == nil {
+		runnerRegexName = regexp.MustCompile(`(?m)([\w -:_]*)\/\[runner.[\w:]*](\/[\w -:_]*)?`)
+	}
+	match := runnerRegexName.FindStringSubmatch(name)
+	if match == nil || len(match) == 0 {
+		return name
+	}
+	return match[1] + match[2]
+}
+
+// Runs the test suite
+func (r *testRunner) Run() int {
+	r.exitCode = 0
+	r.m.Run()
+	return r.exitCode
+}
+
+// Internal test processor, each test calls the processor in order to handle retries, process exiting and exitcodes
+func (r *testRunner) testProcessor(t *testing.T) {
+	t.Helper()
+	name := t.Name()
+	if item, ok := (*r.tests)[name]; ok {
+
+		//Sets the original test name
+		if pointer, err := getFieldPointerOfT(t, "name"); err == nil {
+			*(*string)(pointer) = item.test.Name
+		}
+
+		run := 1
+		maxRetries := r.failRetriesCount
+		exitOnError := r.exitOnError
+		var rc interface{}
+
+		for {
+			var innerTest *testing.T
+			title := "Run"
+			if run > 1 {
+				title = "Retry:" + strconv.Itoa(run-1)
+			}
+			title = "[runner." + title + "]"
+			t.Run(title, func(it *testing.T) {
+				defer func() {
+					rc = recover()
+					if rc != nil {
+						it.FailNow()
+					}
+				}()
+				it.Helper()
+				innerTest = it
+				item.test.F(it)
+			})
+			innerTestInfo := r.getTestResultsInfo(innerTest)
+
+			if rc != nil {
+				if exitOnError {
+					panic(rc)
+				}
+				fmt.Println("PANIC RECOVER:", rc)
+				item.error = true
+			}
+			item.skipped = innerTestInfo.skipped
+			item.ran++
+
+			// Current run failure
+			if innerTestInfo.failed {
+				item.failed = true
+			}
+			// Current run ok but previous with fail -> Flaky
+			if !innerTestInfo.failed && item.failed {
+				item.failed = false
+				item.flaky = true
+				break
+			}
+
+			if item.skipped || !item.failed || run > maxRetries {
+				break
+			}
+			run++
+		}
+		if item.flaky {
+			fmt.Println("*** FLAKY", item.fqn)
+		}
+		if item.error && exitOnError {
+			panic(rc)
+		}
+		if item.error || item.failed {
+			r.exitCode = 1
+		} else {
+			if ptr, err := getFieldPointerOfT(t, "failed"); err == nil {
+				*(*bool)(ptr) = false
+			}
+		}
+	} else {
+		t.FailNow()
+	}
+}
+
+// Initialize test runner
+func (r *testRunner) init() {
+	tests := make([]testing.InternalTest, 0)
+	benchmarks := make([]testing.InternalBenchmark, 0)
+
+	if tPointer, err := r.getFieldPointerOfM("tests"); err == nil {
+		r.intTests = (*[]testing.InternalTest)(tPointer)
+		r.tests = &map[string]*testDescriptor{}
+		for _, test := range *r.intTests {
+			fqn := test.Name
+			(*r.tests)[fqn] = &testDescriptor{
+				test:   test,
+				fqn:    fqn,
+				ran:    0,
+				failed: false,
+			}
+			tests = append(tests, testing.InternalTest{
+				Name: fqn,
+				F:    r.testProcessor,
+			})
```

Comment:
Instead of having a "parallel" data structure `r.tests` to keep metadata about each test, why not wrap each test into an object, that has a method to "run" it? (basically the `testProcessor` method). I think it will be cleaner and easier to read.

Reply:
done